File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        ds = []
        n = len(candidates)
        candidates.sort()

        def rec(ind, target, ds):
            if target == 0:
                ans.append(ds[:])  # Append a copy of the current combination
                return
            for i in range(ind, n):
                if i > ind and candidates[i] == candidates[i - 1]:
                    continue  # Skip duplicates
                if candidates[i] > target:
                    break
                ds.append(candidates[i])
                rec(i + 1, target - candidates[i], ds)
                ds.pop()

        rec(0, target, ds)
        return ans

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    void findcomb(vector<int>& arr, int ind, int target, vector<int>& ds, vector<vector<int>>& comb) {
        if(target == 0){
            comb.push_back(ds);
            return;
        }
        for(int i=ind;i<arr.size();i++){
            if(i>ind && arr[i]==arr[i-1]) continue;
            if(arr[i] >target) break;
            ds.push_back(arr[i]);
            findcomb(arr, i + 1, target - arr[i], ds, comb);
            ds.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> comb;
        vector<int> ds;
        sort(candidates.begin(), candidates.end());
        findcomb(candidates, 0, target, ds, comb);
        return comb;
    }
};
